core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java

```diff
@@ -3053,6 +3054,121 @@ public void testRewriteNegativeLimitWithHiLimitHandlesWildcardsTimestampNotFirst
         });
     }
 
+    @Test
+    public void testRewriteTrivialExpressions() throws Exception {
+        testRewriteTrivialExpressions(false);
+    }
+
+    @Test
+    public void testRewriteTrivialExpressionsInOuterJoin() throws Exception {
+        assertMemoryLeak(() -> {
+            execute(
+                    "CREATE TABLE x (\n" +
+                            "    id long,\n" +
+                            "    ts timestamp\n" +
+                            ") TIMESTAMP(ts) PARTITION BY DAY;"
+            );
+            execute("INSERT INTO x VALUES(1, '2021-01-01T12:34:00');");
+            execute("INSERT INTO x VALUES(2, '2021-01-01T12:35:00');");
+
+            final String query = "WITH x1 AS ( " +
```

Comment:
if i change this SQL to 


```
            final String query = "WITH x1 AS ( " +
                    "  SELECT id id0, id / 2 id1, id / 2 + 1 id2, count(*) AS c " +
                    "  FROM x " +
                    "  GROUP BY id, id / 2 " +
                    "  ORDER BY c DESC" +
                    ") " +
                    "SELECT * " +
                    "FROM x x0 " +
                    "JOIN x1 ON x0.id = x1.id0;";

```

query fails with a weird error

Reply:
i think expression check is superficial, instead it should be recursive. This error could be a net result of incorrect expression check

Reply:
> query fails with a weird error

The query is incorrect as GROUP BY doesn't include the `id / 2 + 1 id2` column. What should this query do?

Reply:
@puzpuzpuz `id / 2` is a derivative of `id` and so is `id / 2 + 1`. Yes, i missed the extra group-by clause, but if you add it you will see we're keying the map on 2 keys instead of one.

The point i was trying to make that logic that analyses trivial expression is superficial right now, it is surface level but it should be recursive instead

Reply:
I understand now. Improved the trivial expression detection in: 99f06df87e0846f3ce8df19c61bceb187e64f73a